tests/test_wasm_intrinsics_simd.c

```diff
@@ -0,0 +1,1476 @@
+v128_t TESTFN i32x4_ge_s(v128_t x, v128_t y) {
+  return wasm_i32x4_ge(x, y);
+}
+v128_t  TESTFN i32x4_ge_u(v128_t x, v128_t y) {
+  return wasm_u32x4_ge(x, y);
+}
+v128_t TESTFN f32x4_eq(v128_t x, v128_t y) {
+  return wasm_f32x4_eq(x,y);
+}
+v128_t TESTFN f32x4_ne(v128_t x, v128_t y) {
+  return wasm_f32x4_ne(x, y);
+}
+v128_t TESTFN f32x4_lt(v128_t x, v128_t y) {
+  return wasm_f32x4_lt(x, y);
+}
+v128_t TESTFN f32x4_gt(v128_t x, v128_t y) {
+  return wasm_f32x4_gt(x,y);
+}
+v128_t TESTFN f32x4_le(v128_t x, v128_t y) {
+  return wasm_f32x4_le(x, y);
+}
+v128_t TESTFN f32x4_ge(v128_t x, v128_t y) {
+  return wasm_f32x4_ge(x, y);
+}
+
+#ifdef __wasm_undefined_simd128__
+
+v128_t TESTFN f64x2_eq(v128_t x, v128_t y) {
+  return wasm_f64x2_eq(x,y);
+}
+v128_t TESTFN f64x2_ne(v128_t x, v128_t y) {
+  return wasm_f64x2_ne(x,y);
+}
+v128_t TESTFN f64x2_lt(v128_t x, v128_t y) {
+  return wasm_f64x2_lt(x,y);
+}
+v128_t TESTFN f64x2_gt(v128_t x, v128_t y) {
+  return wasm_f64x2_gt(x, y);
+}
+v128_t TESTFN f64x2_le(v128_t x, v128_t y) {
+  return wasm_f64x2_le(x, y);
+}
+v128_t TESTFN f64x2_ge(v128_t x, v128_t y) {
+  return wasm_f64x2_ge(x, y);
+}
+
+#endif // __wasm_undefined_simd128__
+
+v128_t TESTFN v128_not(v128_t vec) {
+  return wasm_v128_not(vec);
+}
+v128_t TESTFN v128_and(v128_t x, v128_t y) {
+  return wasm_v128_and(x, y);
+}
+v128_t TESTFN v128_or(v128_t x, v128_t y) {
+  return wasm_v128_or(x,y);
+}
+v128_t TESTFN v128_xor(v128_t x, v128_t y) {
+  return wasm_v128_xor(x,y);
+}
+v128_t TESTFN v128_bitselect(v128_t x, v128_t y, v128_t cond) {
+  return wasm_v128_bitselect(x, y, cond);
+}
+v128_t TESTFN i8x16_neg(v128_t vec) {
+  return wasm_i8x16_neg(vec);
+}
+int32_t TESTFN i8x16_any_true(v128_t vec) {
+  return wasm_i8x16_any_true(vec);
+}
+int32_t TESTFN i8x16_all_true(v128_t vec) {
+  return wasm_i8x16_all_true(vec);
+}
+v128_t TESTFN i8x16_shl(v128_t vec, int32_t shift) {
+  return wasm_i8x16_shl(vec, shift);
+}
+v128_t TESTFN i8x16_shr_s(v128_t vec, int32_t shift) {
+  return wasm_i8x16_shr(vec, shift);
+}
+v128_t TESTFN i8x16_shr_u(v128_t vec, int32_t shift) {
+  return wasm_u8x16_shr(vec, shift);
+}
+v128_t TESTFN i8x16_add(v128_t x, v128_t y) {
+  return wasm_i8x16_add(x,y);
+}
+v128_t TESTFN i8x16_add_saturate_s(v128_t x, v128_t y) {
+  return wasm_i8x16_add_saturate(x, y);
+}
+v128_t TESTFN i8x16_add_saturate_u(v128_t x, v128_t y) {
+  return wasm_u8x16_add_saturate(x, y);
+}
+v128_t TESTFN i8x16_sub(v128_t x, v128_t y) {
+  return wasm_i8x16_sub(x,y);
+}
+v128_t TESTFN i8x16_sub_saturate_s(v128_t x, v128_t y) {
+  return wasm_i8x16_sub_saturate(x, y);
+}
+v128_t TESTFN i8x16_sub_saturate_u(v128_t x, v128_t y) {
+  return wasm_u8x16_sub_saturate(x, y);
+}
+v128_t TESTFN i8x16_mul(v128_t x, v128_t y) {
+  return wasm_i8x16_mul(x, y);
+}
+v128_t TESTFN i16x8_neg(v128_t vec) {
+  return wasm_i16x8_neg(vec);
+}
+bool TESTFN i16x8_any_true(v128_t vec) {
+  return wasm_i16x8_any_true(vec);
+}
+int32_t TESTFN i16x8_all_true(v128_t vec) {
```

Comment:
Should it return bool type as for consistence with other any_true/all_true functions?

Reply:
> Should it return bool type as for consistence with other any_true/all_true functions?

However, if we return bool type here, the compiler will generate a compare to get a Boolean value. Maybe int32_t is the right one or the compiler should try to optimize this. For example
```
012c21 func[150] <i16x8_any_true(int vector[4])>:
 012c22: 20 00                      | local.get 0
 012c24: fd 63                      | i16x8.any_true
 012c26: 41 00                      | i32.const 0
 012c28: 47                         | i32.ne
 012c29: 0b                         | end
```

Reply:
> Should it return bool type as for consistence with other any_true/all_true functions?

Which other `any_true` / `all_true` functions? 

---

> However, if we return bool type here, the compiler will generate a compare to get a Boolean value.

@sunfishcode @tlively this looks like a codegen bug:  `i16x8.any_true` returns an `i32` already (C++ `bool` lowers to an `i32` in WASM), and the `i32.const 0 i32.ne` is unnecessary.

Reply:
Currently the LLVM backend is configured to use `ZeroOrOneBooleanContent` which means that all bools are fixed to be either zero or one. The issue is that LLVM doesn't know that the i32 returned by these WebAssembly instructions is a 0 or a 1 anyway, so it thinks it needs to insert the `and 1`. I [previously](https://reviews.llvm.org/D53676) tried to optimize this out, but that resulted in a [nasty bug](https://bugs.llvm.org/show_bug.cgi?id=40805). I can take a look at optimizing that codegen again, but I suggest that for the sake of this discussion we pretend this codegen issue does not exist.

Reply:
> > Should it return bool type as for consistence with other any_true/all_true functions?
> 
> Which other `any_true` / `all_true` functions?
> 
I mean functions like `bool TESTFN i64x2_any_true` and `bool TESTFN i16x8_any_true`. The return types are not consistent. Some return `bool` and some return `int32_t`.